File: apps/sdkwork-terminal-h5/src/bootstrap/iamRuntime.ts
import { createClient, type SdkworkAppClient } from '@sdkwork/iam-app-sdk';
import {
  createTokenManager,
  type AuthTokenManager,
  type AuthTokens,
} from '@sdkwork/sdk-common';

import { getPlatformApiGatewayHttpUrl } from './environment';
import {
  clearTerminalSession,
  commitTerminalSession,
  readTerminalSession,
  type TerminalSessionSnapshot,
} from './session';

const APP_API_PREFIX = '/app/v3/api';

export interface IamRuntime {
  baseUrl: string;
  tokenManager: AuthTokenManager;
  appbaseApp: SdkworkAppClient;
}

let cachedIamRuntime: IamRuntime | null = null;

export function invalidateIamRuntime(): void {
  cachedIamRuntime = null;
}

function normalizeGeneratedSdkBaseUrl(baseUrl: string, apiPrefix: string): string {
  const normalizedBaseUrl = baseUrl.replace(/\/+$/, '');
  const normalizedApiPrefix = apiPrefix.replace(/\/+$/, '');
  if (normalizedBaseUrl.endsWith(normalizedApiPrefix)) {
    return normalizedBaseUrl.slice(0, -normalizedApiPrefix.length) || normalizedBaseUrl;
  }
  return normalizedBaseUrl;
}

function hydrateTokenManager(
  tokenManager: AuthTokenManager,
  snapshot: TerminalSessionSnapshot | null,
): void {
  if (!snapshot) {
    return;
  }

  tokenManager.setTokens({
    ...(snapshot.accessToken ? { accessToken: snapshot.accessToken } : {}),
    ...(snapshot.authToken ? { authToken: snapshot.authToken } : {}),
    ...(snapshot.refreshToken ? { refreshToken: snapshot.refreshToken } : {}),
  });
}

function bindSessionPersistence(tokenManager: AuthTokenManager): void {
  const originalSetTokens = tokenManager.setTokens.bind(tokenManager);
  const originalClearTokens = tokenManager.clearTokens.bind(tokenManager);

  tokenManager.setTokens = (tokens: AuthTokens) => {
    const result = originalSetTokens(tokens);
    commitTerminalSession({
      accessToken: tokens.accessToken,
      authToken: tokens.authToken,
      refreshToken: tokens.refreshToken,
    });
    return result;
  };

  tokenManager.clearTokens = () => {
    originalClearTokens();
    clearTerminalSession();
  };
}

function createAppbaseAppClient(tokenManager: AuthTokenManager): SdkworkAppClient {
  return createClient({
    authMode: 'dual-token',
    baseUrl: normalizeGeneratedSdkBaseUrl(
      getPlatformApiGatewayHttpUrl(),
      APP_API_PREFIX,
    ),
    platform: 'h5',
    tokenManager,
  });
}

export function createIamRuntime(): IamRuntime {
  const tokenManager = createTokenManager();
  bindSessionPersistence(tokenManager);
  hydrateTokenManager(tokenManager, readTerminalSession());

  return {
    baseUrl: getPlatformApiGatewayHttpUrl(),
    tokenManager,
    appbaseApp: createAppbaseAppClient(tokenManager),
  };
}

export function getIamRuntime(): IamRuntime {
  if (!cachedIamRuntime) {
    cachedIamRuntime = createIamRuntime();
  }
  return cachedIamRuntime;
}

export type TokenManager = AuthTokenManager;

File: apps/sdkwork-terminal-flutter-mobile/lib/bootstrap/iam_runtime.dart
import 'package:sdkwork_appbase_app_sdk/sdkwork_appbase_app_sdk.dart';

import 'environment.dart';
import 'session.dart';

const _appApiPrefix = '/app/v3/api';

class IamRuntime {
  final String baseUrl;
  final TerminalSessionSnapshot? session;
  final SdkworkAppClient appbaseApp;

  IamRuntime._({
    required this.baseUrl,
    required this.session,
    required this.appbaseApp,
  });

  bool get hasIamSession => session?.hasIamSession ?? false;

  factory IamRuntime.create({TerminalSessionSnapshot? session}) {
    final baseUrl = _normalizeGeneratedSdkBaseUrl(
      Environment.platformApiGatewayHttpUrl,
      _appApiPrefix,
    );
    final resolvedSession = session;

    return IamRuntime._(
      baseUrl: baseUrl,
      session: resolvedSession,
      appbaseApp: SdkworkAppClient.withBaseUrl(
        baseUrl: baseUrl,
        authToken: resolvedSession?.authToken,
        accessToken: resolvedSession?.accessToken,
      ),
    );
  }

  static String _normalizeGeneratedSdkBaseUrl(String baseUrl, String apiPrefix) {
    final normalizedBaseUrl = baseUrl.replaceAll(RegExp(r'/+$'), '');
    final normalizedApiPrefix = apiPrefix.replaceAll(RegExp(r'/+$'), '');
    if (normalizedBaseUrl.endsWith(normalizedApiPrefix)) {
      final trimmed = normalizedBaseUrl.substring(
        0,
        normalizedBaseUrl.length - normalizedApiPrefix.length,
      );
      return trimmed.isEmpty ? normalizedBaseUrl : trimmed;
    }
    return normalizedBaseUrl;
  }
}

IamRuntime? _cachedIamRuntime;

void initializeIamRuntime(TerminalSessionSnapshot? session) {
  _cachedIamRuntime = IamRuntime.create(session: session);
}

void invalidateIamRuntime() {
  _cachedIamRuntime = null;
}

IamRuntime getIamRuntime() {
  return _cachedIamRuntime ??= IamRuntime.create();
}
